### PR TITLE
Add support for ElasticStack Built-In  Users

### PR DESCRIPTION
**Pull request by raihankhan**
Signed-off-by: raihankhan <raihan@appscode.com>

### CODE REVIEW

File: pkg/util/go-es/es_client_v6.go [hunk 1]
```diff
@@ -48,3 +49,7 @@ func (es *ESClientV6) ClusterStatus() (string, error) {
 	}
 	return "", errors.New("status is missing")
 }
+
+func (es *ESClientV6) SyncCredentialFromSecret(secret *core.Secret) error {
```
Comment (kamolhasan) [inline pkg/util/go-es/es_client_v6.go:53]: Add comments

File: pkg/util/go-es/es_client_v7.go [hunk 2]
```diff
@@ -48,3 +55,54 @@ func (es *ESClientV7) ClusterStatus() (string, error) {
...
+	}
+
+	defer func(Body io.ReadCloser) {
+		err = Body.Close()
+		if err != nil {
+			klog.Errorf("failed to close auth response body", err)
+		}
+	}(res.Body)
+
+	if res.StatusCode == http.StatusOK {
```
Comment (kamolhasan) [inline pkg/util/go-es/es_client_v7.go:101]: Use isError()

**Review by kamolhasan — CHANGES_REQUESTED** [review]

File: pkg/admission/validator.go [hunk 3]
```diff
@@ -238,6 +249,22 @@ func ValidateElasticsearch(client kubernetes.Interface, extClient cs.Interface,
...
+			return nil
+		}
+		if version.Major() < 7 {
+			return fmt.Errorf(`'spec.internalUsers' is unsupported for version < 7`)
+		}
+	}
+	// Allow support for only builtin internal users
+	// custom users can not be created via client api request
+	// ref: https://github.com/elastic/go-elasticsearch/find/main
+	if err := amv.ValidateInternalUsers(db.Spec.InternalUsers, allowedInternalUsers, api.ResourceKindElasticsearch); err != nil {
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/admission/validator.go:264]: done

File: pkg/controller/appbinding.go [hunk 4]
```diff
@@ -75,6 +75,25 @@ func (c *Controller) ensureAppBinding(db *api.Elasticsearch) (kutil.VerbType, er
 		} else {
 			caBundle = v
 		}
+
+		// ensure and validate client-cert before referencing in AppBinding
+		sName, exist = api_util.GetCertificateSecretName(db.Spec.TLS.Certificates, string(api.ElasticsearchClientCert))
+		if !exist {
+			return kutil.VerbUnchanged, errors.New("client-cert secret is missing")
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/controller/appbinding.go:82]: Changed

File: pkg/controller/appbinding.go [hunk 5]
```diff
@@ -75,6 +75,25 @@ func (c *Controller) ensureAppBinding(db *api.Elasticsearch) (kutil.VerbType, er
...
+		}
+
+		clientCertSecret, err := c.Client.CoreV1().Secrets(db.Namespace).Get(context.TODO(), sName, metav1.GetOptions{})
+		if err != nil {
+			return kutil.VerbUnchanged, errors.Wrapf(err, "failed to read client-cert secret for Elasticsearch %s/%s", db.Namespace, db.Name)
+		}
+
+		if _, ok := clientCertSecret.Data[certlib.CACert]; !ok {
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/controller/appbinding.go:90]: removed redundant check

File: pkg/controller/elasticsearch.go [hunk 6]
```diff
@@ -151,26 +152,61 @@ func (c *Controller) create(db *api.Elasticsearch) error {
...
+		// Get original Elasticsearch version, since the client is version specific
+		esVersion, err := c.DBClient.CatalogV1alpha1().ElasticsearchVersions().Get(context.TODO(), db.Spec.Version, metav1.GetOptions{})
+		if err != nil {
+			return errors.Wrap(err, "failed to get elasticsearchVersion")
+		}
+		version, err := semver.NewVersion(esVersion.Spec.Version)
 		if err != nil {
 			return err
 		}
+		if (version.Major() == 7 && version.Minor() >= 8) || version.Major() >= 8 {
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/controller/elasticsearch.go:165]: fixed

File: pkg/util/go-es/es_client_v7.go [hunk 7]
```diff
@@ -17,19 +17,26 @@ limitations under the License.
...
 	esv7 "github.com/elastic/go-elasticsearch/v7"
+	"github.com/elastic/go-elasticsearch/v7/esapi"
 	"github.com/pkg/errors"
+	core "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
 )
 
 type ESClientV7 struct {
-	client *esv7.Client
+	AdminClient *esv7.Client
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/util/go-es/es_client_v7.go:34]: Fixed

File: pkg/util/go-es/api.go [hunk 8]
```diff
@@ -41,14 +43,16 @@ import (
 
 type ESClient interface {
 	ClusterStatus() (string, error)
+	SyncCredentialFromSecret(secret *core.Secret) error
 }
 
 func GetElasticClient(kc kubernetes.Interface, db *api.Elasticsearch, esVersion *catalog.ElasticsearchVersion, url string) (ESClient, error) {
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/util/go-es/api.go:49]: Added

File: pkg/distribution/elastic_stack/credential.go [hunk 9]
```diff
@@ -121,6 +228,120 @@ func (es *Elasticsearch) createOrSyncAdminCredSecret() (*core.LocalObjectReferen
...
+		// so, if db is not ready or db not accepting connection abort syncing
+		if !kmapi.IsConditionTrue(es.db.Status.Conditions, api.DatabaseAcceptingConnection) ||
+			!kmapi.IsConditionTrue(es.db.Status.Conditions, api.DatabaseReplicaReady) {
+			return internalUsers, nil
+		}
+		err = es.validateAndSyncLabels(secret, user)
+		if err != nil {
+			return internalUsers, errors.Wrap(err, "failed to validate and sync Labels")
+		}
+		url := fmt.Sprintf("%v://%s.%s.svc:%d", es.db.GetConnectionScheme(), es.db.ServiceName(), es.db.GetNamespace(), api.ElasticsearchRestPort)
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/distribution/elastic_stack/credential.go:273]: Added

File: pkg/distribution/elastic_stack/credential.go [hunk 10]
```diff
@@ -121,6 +228,120 @@ func (es *Elasticsearch) createOrSyncAdminCredSecret() (*core.LocalObjectReferen
 	return es.db.Spec.AuthSecret, nil
 }
 
+// creates secret for internalUsers if not exists
+// if secret exists, sync with the credentials
+// if sync fails, patch internalUsers condition to false
+// and return error
+func (es *Elasticsearch) createOrSyncBuiltinUserCred(user string) (map[string]api.ElasticsearchUserSpec, error) {
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/distribution/elastic_stack/credential.go:235]: FIxed

File: pkg/distribution/elastic_stack/credential.go [hunk 11]
```diff
@@ -64,6 +70,107 @@ func (es *Elasticsearch) EnsureAuthSecret() error {
...
+	internalUsers, err := es.createOrSyncBuiltinUserCred(string(api.ElasticsearchInternalUserKibanaSystem))
+	if err != nil {
+		return err
+	}
+	_, _, err = util.CreateOrPatchElasticsearch(context.TODO(), es.extClient.KubedbV1alpha2(), es.db.ObjectMeta, func(in *api.Elasticsearch) *api.Elasticsearch {
+		in.Spec.InternalUsers = internalUsers
+		return in
+	}, metav1.PatchOptions{})
+	if err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/distribution/elastic_stack/credential.go:98]: created

File: pkg/distribution/elastic_stack/credential.go [hunk 12]
```diff
@@ -64,6 +70,107 @@ func (es *Elasticsearch) EnsureAuthSecret() error {
...
+	}
+	if version.Major() < 7 || (version.Major() == 7 && version.Minor() < 8) {
+		return nil
+	}
+
+	// The user Kibana uses to connect and communicate with Elasticsearch
+	// elastic user is deprecated for kibana for versions 8.x.x
+	internalUsers, err := es.createOrSyncBuiltinUserCred(string(api.ElasticsearchInternalUserKibanaSystem))
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/distribution/elastic_stack/credential.go:89]: Algorithm modified

File: pkg/util/go-es/es_client_v7.go [hunk 13]
```diff
@@ -48,3 +55,54 @@ func (es *ESClientV7) ClusterStatus() (string, error) {
...
+	}
+	if value, ok := secret.Data[core.BasicAuthPasswordKey]; ok {
+		password = string(value)
+	} else {
+		return errors.New("password is missing")
+	}
+
+	// Build the request body.
+	var b strings.Builder
+	b.WriteString(`{"password" : "`)
```
[the remark this answers is not in the document]
Comment (raihankhan) [reply pkg/util/go-es/es_client_v7.go:75]: Fixed

File: pkg/util/go-es/es_client_v7.go (hunk 2 above)
Comment (raihankhan) [reply pkg/util/go-es/es_client_v7.go:101]: Fixed

File: pkg/util/go-es/api.go [hunk 14]
```diff
@@ -41,38 +43,42 @@ import (
...
 	var username, password string
-	if !db.Spec.DisableSecurity && db.Spec.AuthSecret != nil {
-		secret, err := kc.CoreV1().Secrets(db.Namespace).Get(context.TODO(), db.Spec.AuthSecret.Name, metav1.GetOptions{})
-		if err != nil {
-			klog.Errorf("Failed to get secret: %s for Elasticsearch: %s/%s with: %s", db.Spec.AuthSecret.Name, db.Namespace, db.Name, err.Error())
-			return nil, errors.Wrap(err, "failed to get the secret")
-		}
 
-		if value, ok := secret.Data[core.BasicAuthUsernameKey]; ok {
+	if !opt.Db.Spec.DisableSecurity && opt.Db.Spec.AuthSecret != nil {
```
Comment (kamolhasan) [inline pkg/util/go-es/api.go:64]: Check whether the opt.Secret is empty or not?

File: pkg/util/go-es/api.go [hunk 15]
```diff
@@ -41,38 +43,42 @@ import (
 
 type ESClient interface {
 	ClusterStatus() (string, error)
+	SyncCredentialFromSecret(secret *core.Secret) error
 }
 
-func GetElasticClient(kc kubernetes.Interface, db *api.Elasticsearch, esVersion *catalog.ElasticsearchVersion, url string) (ESClient, error) {
-	if db == nil || esVersion == nil {
+type ClientOptions struct {
+	Kc        kubernetes.Interface
```
Comment (kamolhasan) [inline pkg/util/go-es/api.go:50]: KClient
DB
ESVersion
URL

File: pkg/distribution/elastic_stack/credential.go [hunk 16]
```diff
@@ -121,6 +132,247 @@ func (es *Elasticsearch) createOrSyncAdminCredSecret() (*core.LocalObjectReferen
...
+	}
+
+	// The user Kibana uses to connect and communicate with Elasticsearch
+	// elastic user is deprecated for kibana for versions 8.x.x
+	secret, err := es.createOrSyncBuiltinUserCred(api.ElasticsearchInternalUserKibanaSystem)
+	if err != nil {
+		return err
+	}
+	_, _, err = util.CreateOrPatchElasticsearch(context.TODO(), es.extClient.KubedbV1alpha2(), es.db.ObjectMeta, func(in *api.Elasticsearch) *api.Elasticsearch {
+		if user, ok := in.Spec.InternalUsers[string(api.ElasticsearchInternalUserKibanaSystem)]; ok {
```
Comment (kamolhasan) [inline pkg/distribution/elastic_stack/credential.go:279]: check, if map is nil. 
Then use make(map[]) to initialize.

**Review by kamolhasan — CHANGES_REQUESTED** [review]

**Review by kamolhasan — APPROVED** [review]
LGTM!

**Pull request merged**